### PR TITLE
feat(3d from 4d): 3D image generation from 4D

### PR DESCRIPTION
**Pull request by NeilMacPhee**
### Summary
This feature will allow users to choose between three methods to generate a 3D volume from 4D data: subtract, average, or sum. The "subtract" method works with two time frames and will subtract the values of each voxel in the 4D data from the corresponding voxel in the previous frame. The "average" method will take the average of the values of each voxel in the 4D data over time. The "sum" method will add the values of each voxel in the 4D data over time. The output scalar array can then be used to create a 3D volume.

"generateImageFromTimeData" allows the user to specify which time frames to use for generating the scalar data.

### Demo

3DimageGeneration

### CODE REVIEW

File: packages/tools/src/utilities/dynamicVolume/generateImageFromTimeData.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  const arrayLength = typedArrays[0].length;
+  const finalArray = new Float32Array(arrayLength);
+
+  if (operation === 'SUM') {
+    for (let i = 0; i < numFrames; i++) {
+      const currentArray = typedArrays[frames[i]];
+      for (let j = 0; j < arrayLength; j++) {
+        finalArray[j] += currentArray[j];
+      }
+    }
```
Comment (sedghi) [inline packages/tools/src/utilities/dynamicVolume/generateImageFromTimeData.ts:40]: you can return final array at the end of this

File: packages/tools/src/utilities/dynamicVolume/generateImageFromTimeData.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    throw new Error('Please provide two or more time points');
+  }
+
+  // Gets scalar data for all time frames
+  const typedArrays = dynamicVolume.getScalarDataArrays();
+
+  const arrayLength = typedArrays[0].length;
+  const finalArray = new Float32Array(arrayLength);
+
+  if (operation === 'SUM') {
```
Comment (sedghi) [inline packages/tools/src/utilities/dynamicVolume/generateImageFromTimeData.ts:34]: operations should be Enums that you use, similar to viewportTypes etc

File: packages/tools/src/utilities/dynamicVolume/generateImageFromTimeData.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+      }
+    }
+  }
+
+  if (operation === 'SUBTRACT') {
+    if (frames.length > 2) {
+      throw new Error('Please provide only 2 time points for subtraction.');
+    }
+    for (let j = 0; j < arrayLength; j++) {
+      finalArray[j] += typedArrays[0][j] - typedArrays[1][j];
```
Comment (sedghi) [inline packages/tools/src/utilities/dynamicVolume/generateImageFromTimeData.ts:48]: why subtracting 0 and 1? it should accept frames

**Review by sedghi — COMMENTED** [review]
see my comments

File: packages/tools/src/utilities/dynamicVolume/generateImageFromTimeData.ts (hunk 1 above)
Comment (NeilMacPhee) [reply packages/tools/src/utilities/dynamicVolume/generateImageFromTimeData.ts:41]: Moved 'return finalArray' to the end of each operator function

File: packages/tools/src/utilities/dynamicVolume/generateImageFromTimeData.ts (hunk 2 above)
Comment (NeilMacPhee) [reply packages/tools/src/utilities/dynamicVolume/generateImageFromTimeData.ts]: Added Enums DynamicOperatorType, changed if statements to use types

File: packages/tools/src/utilities/dynamicVolume/generateImageFromTimeData.ts (hunk 3 above)
Comment (NeilMacPhee) [reply packages/tools/src/utilities/dynamicVolume/generateImageFromTimeData.ts]: changed to typedArrays[frames[0]][j] - typedArrays[frames[1]][j], was using this during development and forgot to change it back

**Review by sedghi — APPROVED** [review]
Looks great. This is a great contribution, thanks a lot!

**Pull request merged**